### ISSUE TITLE
Use TPM for long-term identity keys

### ISSUE THREAD

**Issue by drbild** [xaptum/enftun#124]
Add support for generating the long-term ECDSA identity key pair in the TPM.

### High-Level Tasks
- Provisioning
  - XTT Flow
    - [x] Update xaptum/xtt to support generating the key and cert using the TPM
  - enftun-keygen Flow
    - [x] Update `enftun-keygen` to support generating the key and cert using the TPM
- Operation
  - [x] Update `enftun` to support using key in the TPM

###  Design Decisions

I see two primary design decisions to be made.  How is the private key "stored" and referenced for `enftun` to use.  Is the certificate stored during provisioning or recreated on each use?

#### Where is the Private Key Stored?

Obviously, the actual private key never leaves the TPM.  However, some identifier must be stored to allow it to be reference and used.

1) Transient keys with File-based PEM

    The most commonly used approach today (gnutls, the openSSL TPM and TPM2 engines, openconnect) is the PEM-wrapped key [1], [2], [3], [4], [5].

    ```
    -----BEGIN TSS2 PRIVATE KEY ---
    <ASN.1 OCTET STREAM>
    -----END TSS2 PRIVATE KEY-----
    ```

    [1] https://github.com/tpm2-software/tpm2-tools/issues/1599
    [2] https://lists.ofono.org/hyperkitty/list/tpm2@lists.01.org/thread/3O4CQKKRXFTUDSHADTGRYWHFJFWJBLYA/
    [3] Sec 4.4, Paragraph 3: http://david.woodhou.se/draft-woodhouse-cert-best-practice.html
    [4] https://blog.hansenpartnership.com/using-your-tpm-as-a-secure-key-store/
    [5] https://mta.openssl.org/pipermail/openssl-dev/2016-November/008839.html

    This approach has several benefits:
    - Fits cleanly with the non-TPM flow.  We still store a private key file (and `enftun` still loads that private key file), it just contains a reference to the key, instead of the entire key. ( For now, `enftun` would be required to parse the PEM header (`BEGIN ...`) to determine whether to load the key with the TSS engine or the traditiional methods.  But eventually, openSSL will be taught that header and will be able to invoke the correct engine automatically.)
    - Doesn't require persisting the private key
    - No changes to the openSSL `tpm2-tss-engine`
    - Easily support an arbitrary number of `enftun` "instances" on one TPM. 
 
   And one potential con:
   - It requires storing something to disk.  We currently require that, but having the option to support a system without writeable non-volatile storage might be nice in the future.  Of course, this is complicated by the need to store the device IPv6 identity/certificate, so isn't a high priority right now. 

2) Persistent keys with Handles

   We could instead persist the key into the TPM NVRAM storage.  This will require just storing a handle to the key.

   If we only need to support on `enftun` "instance" per TPM, the handle can be hard-coded and not explicitly stored on the disk.  But the router cards need two instances, so I think the handle still needs to be stored.

Given that the PEM-based approach is common in the ecosystem, I will advocate for it. 

EDIT: Based on my new findings in the comments below, it's looks like both approaches are equally supported by the TPM2 engines.  I'm now going to advocate that we not use an engine for key creation (in either `enftun-keygen` or `xtt`) to avoid a dependency on a specific engine. Since we'll code that ourselves, we can support PEM wrapped file, persistent key handle, or both.

#### When is the Certificate Generated

The current flow is to generate, sign, and store the certificate during provisioning.   The cert could be regenerated/signed on each use, to eliminate the need to store it.  However, the certificate is the only place where the device identity (IPv6) is stored.   So eliminating storage of the certificate would add storage of the device identity.  Doesn't seem worth it.

CC @dberliner @zanebeckwith

**Comment by zanebeckwith** [comment]
The XTT work for storing the longterm key is tracked [here](https://github.com/xaptum/xtt/issues/120)

**Comment by drbild** [comment]
An additional realization:

`enftun` shouldn't (need to be) married to a specific OpenSSL TPM2 engine. 

There are two in development today. The [tpm2-tss-engine](https://github.com/tpm2-software/tpm2-tss-engine) from the tpm2-software group and [openssl-tpm-engine](https://git.kernel.org/pub/scm/linux/kernel/git/jejb/openssl_tpm2_engine.git/) from James Bottomley.

OpenConnect shows how to support both.

In the application:
1. Don't `#include` any engine-specific headers and don't link the application against the engine. Just use the generic openSSL [headers]().
1. Specify `#define OPENSSL_LOAD_CONF` or manually call `OPENSSL_config` to read the local openSSL config at runtime.
1. Use the generic openSSL functions to [load all] configured engines and locate either TPM2 engine ("tpm2" or "tpm2tss").

On the running the application:
1. Install the desired openSSL engine.
1. Configure the global `/etc/ssl/openssl.cnf` to configure the engine.  See examples for [openssl-tpm2-engine](https://git.kernel.org/pub/scm/linux/kernel/git/jejb/openssl_tpm2_engine.git/tree/openssl.cnf.sample) and [tpm2-tss-engine](https://github.com/tpm2-software/tpm2-tss-engine/blob/master/openssl.conf.sample).

With this in mind, I think that we _should not_ rely on `tpm2-tss-engine` functions to create the keys in `enftun-keygen`.  By implementing it outselves, we won't have any dependency on a particular engine. 

Even better, this decoupling ensures compatibility with future versions of the engines.  The openSSL API isn't going to change on us (until openSSL releases a new significant version), but the engine-specific methods for creating keys might.

**Comment by drbild** [comment]
It looks like `tpm2-tss-engine` can support either a handle or a PEM-backed file:

The engine [sets](https://github.com/tpm2-software/tpm2-tss-engine/blob/62a60ef9aa08726afa72dcd5361c4089ff78a263/src/tpm2-tss-engine.c#L191) `loadkey` for the `ENGINE_load_privkey` callback.

`loadkey` [checks](https://github.com/tpm2-software/tpm2-tss-engine/blob/62a60ef9aa08726afa72dcd5361c4089ff78a263/src/tpm2-tss-engine.c#L191) if the supplied key id looks like a handle  (`0x81...`).  If so, it calls `tpm2tss_tpm2data_readtpm` to read the key from the TPM.  Otherwise, it assumes the key id is a file name and calls `tpm2tss_tpm2data_read` to read the key from a PEM-wrapped file.

So from an `enftun` perspective, supplying either a handle `0x81...` or a file name (as is done today) in the `identity.key_file` [config option](https://github.com/xaptum/enftun/blob/master/example/device.conf#L83) should work.

On the provisioning side, `enftun` and `xtt` would need to parse that config option to determine if a persistent handle or file-backed key is desired.

**Comment by zanebeckwith** [comment]
Regarding the storage of the certificate:

Yes, I agree that the cert shouldn't be stored in the TPM (or the identity stored in the TPM and the cert regenerated on demand).

It's not a sensitive file, so there's no security gain from keeping it in the TPM.

And, the only use-case I can think of for keeping it off the host's disk is just a limited host that has no storage. That's an interesting use-case, but not one we're facing now. Adding the logic to store the cert to NVRAM isn't trivial, so I say we put that off for a future enhancement.

By storing the cert on disk, once the key is generated all the steps for cert generation and storage should be exactly the same as in the non-TPM case. The TPM case will simply be chosen by 1) loading the TPM engine and 2) specifying a TPM key as the key to use

**Comment by zanebeckwith** [comment]
Regarding the multiple engines:

Ah, I didn't realize James Bottomley's engine was still active (I actually for some reason thought his project had morphed into the `tpm2-software` one...). So, yea, I agree: with two engines in the wild, keeping ourselves agnostic to the particular one used is definitely a good idea.

To make sure I understand what you're sketching out: the core logic of `enftun` will be agnostic to the engine used, but the `enftun` package will include an OpenSSL config file that specifies a particular one (we can choose whichever we like) and has a dependency on that one. Of course, if a user prefers the other engine, they can install that one and use an override config file.

Is that right?

This project (`enftun`) shouldn't alter the global config file, only its particular config file, right?

**Comment by zanebeckwith** [comment]
Regarding the blob-file vs persistent-handle discussion:

Yes, we were aware that the `tpm2-tss-engine` transparently supports using either a file or handle for `load`ing.

Dan, correct me if I'm wrong, but the plan was to:
- Have logic to generate a persistent key
  - This logic would have to involve some amount of custom work, either doing it all ourselves, or piecing together existing `tpm2-tss-engine` functionality, or upstreaming a "evict-this-key" function to the tpm2-tss-engine project
- Then do all other operations (cert generation, SSL handshake) exactly as usual
  - If the key's in the TPM, the OpenSSL library can use it fine because we specify the key by its handle
  - If the key's just a typical file, the OpenSSL library of course does its usual thing

I hadn't considered storing a TPM key as a file (rather than a handle), but other than that question (do we generate the key as a blob file, or as a persisted key handle?) yes I think we're all on the same page regarding how the rest of the flow will go. I'm still fuzzy on how the engine gets configured (as my other comment shows), but I think I understand at a high level.

My reasoning behind advising Dan to use the `tpm2-tss-engine` keygen stuff as much as possible was a general don't-reinvent-the-wheel effort reduction. But, with two engines available, I agree that keeping ourselves engine-agnostic is a good idea.

David, you've convinced me that storing the key as a blob file is an excellent option. I was stuck in a "this is how we did it for the DAA key" mindset, and didn't really think through that a blob file would work fine for this project.

Long-term, supporting both a blob file *and* a handle for the key is a good idea.

Correct me if I'm wrong, but the steps for generating a persistent key with a handle and the steps for generating a key as a blob file are:

Persistent key:
- Create a primary key
- Create a child key under that primary
- Load the child key
- Evict the child key ("saved" to a provided handle)

Blob-file:
- Create a primary key
- Create a child key under that primary

In other words, saving the key as a blob file requires a subset of the TPM commands that persisting the key as a handle does.

On the other hand, saving the created key as a blob file requires getting together the ASN.1 and PEM tools.

We have examples of both flows (between `tpm2-tss-engine` and `ecdaa`).

My gut is that for the short-term just supporting the blob file may be easiest. David, your comment in the PR mentioned above makes me confident we don't have to worry about the TPM running out of memory and so being out in the cold when we try to load the blob file (we can assume the TPM Resource Manager is helping out there).

**Comment by zanebeckwith** [comment]
An alternative method of generating the key (alternative to either "storing" the key as a blob file or as a persistent handle) is to use a primary key (rather than a child key) and just regenerate it every time we need to use it.

David, you mentioned this option in the original PR that started the conversation in this issue, so I wanted to see that discussion through.

From the viewpoint of our coding effort for the *initial* creation of the key, this is maybe easier than either of the other two options: all we have to do is create a primary key, no need for the other steps

We can ensure each "instance" of enftun has a different key by generating some random bytes at "first" key creation time, using those bytes in the "OutsideInfo" parameter to the CreatePrimary function, then storing those random bytes to file (this file now plays the role of the "key file", i.e. it's necessary when "loading", actually regenerating, the key)

However, this approach has the significant drawback that it now requires extra effort on our part before using the key for typical SSL operations (cert creation or SSL handshake): instead of simply passing the handle or blob-file to the OpenSSL library and letting it load the key for us, we have to do the regeneration each time.

An advantage of this approach is that it gives us the option of *not* using the "OutsideInfo" (and thus not needing to store the random bytes), and so obviating the need to store anything for the key. However, as discussed in the comment above about the cert storage, I don't think that's a significant advantage right now

**Comment by drbild** [comment]
> Ah, I didn't realize James Bottomley's engine was still active (I actually for some reason thought his project had morphed into the `tpm2-software` one...). 

I had as well, until yesterday...

> To make sure I understand what you're sketching out: the core logic of `enftun` will be agnostic to the engine used, but the `enftun` package will include an OpenSSL config file that specifies a particular one (we can choose whichever we like) and has a dependency on that one. Of course, if a user prefers the other engine, they can install that one and use an override config file.
> 
> Is that right?
> 
> This project (`enftun`) shouldn't alter the global config file, only its particular config file, right?

I don't think `enftun` should even provide a config file at all.  We can provide instructions (or links to instructions) for configuring the engine(s), but leave it up to the user to actually do that configuration.

Our packaging for a particular OS (say Debian) could include a recommended dependency on one of the engines. That engine package might configure openSSL properly & automatically.  I don't see `tpm2-tss-engine` [packages](https://repology.org/project/tpm2-tss-engine/versions) for Debian or Ubuntu anyway.  `openssl-tpm2-engine` has an [official package](https://repology.org/project/openssl-tpm2-engine/versions) for openSUSE rolling release, but [unofficial packages](https://build.opensuse.org/package/show/home:jejb1:TPM/openssl_tpm2_engine) for lots of OSs via the Open Build Service.

**Comment by drbild** [comment]
> My reasoning behind advising Dan to use the `tpm2-tss-engine` keygen stuff as much as possible was a general don't-reinvent-the-wheel effort reduction. But, with two engines available, I agree that keeping ourselves engine-agnostic is a good idea.

I wish the OpenSSL engine interface included a "create_keypair" function...

**Comment by zanebeckwith** [comment]
> I wish the OpenSSL engine interface included a "create_keypair" function...

Ha, I had the exact some thought! Dan, I think we went through this on a phone call a while back. "Wait, where does it register the genkey function, for key creation? Can we just use the typical OpenSSL call for key generation, but using this engine in the backend??"

**Comment by drbild** [comment]
> - Create a primary key
> - Create a child key under that primary

> From the viewpoint of our coding effort for the _initial_ creation of the key, this is maybe easier than either of the other two options: all we have to do is create a primary key, no need for the other steps

I'm unclear on the reasons for or against using a custom primary key versus a user-provided one.  I think I just don't understand the hierarchies and ownership model well enough.

The TPM2 [book](https://link.springer.com/content/pdf/10.1007%2F978-1-4302-6584-9.pdf) says that "_In most use cases, at least one storage primary key is made persistent in the TPM for the storage hierarchy, to act in a manner similar to the SRK_".

`tpm2-tss-engine` [uses](https://github.com/tpm2-software/tpm2-tss-engine/blob/62a60ef9aa08726afa72dcd5361c4089ff78a263/src/tpm2-tss-engine-common.c#L464) creates one under `TPM2_RH_OWNER` if an explicit parent handle was not provided.

Multiple instances of `enftun` can share the same primary key, right?  Just different child keys.

Zane, do you have any clarity on the intended organization here?

**Comment by zanebeckwith** [comment]
Because the XTT library will need this same functionality, I'm wondering if we should have this key generation stuff in a new repository.

That project would have a library with at least the following functions:
- `gen_key`
  - This calls CreatePrimary and Create (to create the primary key and the child key), and outputs 1) the `OutPrivate` encrypted private key blob and 2) the `OutPublic` public key
- `write_key`
  - This writes the encrypted private key blob etc. to a PEM-encoded file (I'm still unclear on what exactly gets written in that file; I haven't followed up on it yet)
- `load_key`
  - This takes the `OutPublic` and `OutPrivate` blobs and calls Load, then outputs the handle of the loaded key
  - This probably also needs to regenerate the parent. I need to look at how `tpm2-tools` does this, for example
- `sign`
  - This takes the key handle and the buffer to sign and calls Sign

The first two functions are necessary for both this project and `xtt`, while the second two functions are only necessary for `xtt` (it can't rely on the OpenSSL implementations of these)

It could also provide a binary that basically calls the `gen_key` and `write_key` functions

Originally, I had thought this project would be tied to OpenSSL and we'd want to reuse as much from there as possible, whereas `xtt` would have other concerns (e.g. actually signing something using the key), and so I thought we'd keep these separate. But, if we're implementing all the key generation ourselves, I think we'll need this for both projects. I haven't followed through on how exactly we'd need to structure this other repo, to be useful for both `enftun` and `xtt`, though

**Comment by zanebeckwith** [comment]
> `tpm2-tss-engine` [creates](https://github.com/tpm2-software/tpm2-tss-engine/blob/62a60ef9aa08726afa72dcd5361c4089ff78a263/src/tpm2-tss-engine-common.c#L464) one under `TPM2_RH_OWNER` if an explicit parent handle was not provided.

Right, and my thought was that we'd have the same organization (and possibly just not even allow the user to specify an existing parent key, just *always* creating a new one)

Because a primary key created with the same parameters is always exactly the same (cf. the "regeneration" discussion), "creating" a "new" primary key each time is conceptually the same as using the same key for all of the `enftun` child keys

Now, whether the TPM actually creates a separate primary key (i.e. uses storage) each time you try that, or just creates it if it hasn't already been created, I don't know

**Comment by drbild** [comment]
> Right, and my thought was that we'd have the same organization (and possibly just not even allow the user to specify an existing parent key, just always creating a new one)
>
> Because a primary key created with the same parameters is always exactly the same (cf. the "regeneration" discussion), "creating" a "new" primary key each time is conceptually the same as using the same key for all of the `enftun` child keys

What's the reason to create our own primary, rather than just use the one created by the engine?

Worded more precisely, is there a need to create a _different_ primary key from the one that would be created by the engine?  Or can we just generate the same primary key that the engine will generate, and thus not need to persist it/specify a handle?

**Comment by zanebeckwith** [comment]
> And the engine [uses](https://github.com/tpm2-software/tpm2-tss-engine/blob/62a60ef9aa08726afa72dcd5361c4089ff78a263/src/tpm2-tss-engine-common.c#L453) `Esys_TR_FromTPMPublic` to load the primary key, so I think that means it has to be a persistent key.

I'm confused by this comment. That code is run if you provide a parent handle. So, yes, if you provide a handle for the parent (which doesn't really need to be a primary key, we could be further down a tree of keys), then it needs to be able to use it via a handle

However, if you don't provide a parent handle to that function, it creates it (or, maybe more precisely, recreates it)

**Comment by zanebeckwith** [comment]
> EDIT: (...)

Ah, sorry, now I see what you're saying. Interesting...

**Comment by zanebeckwith** [comment]
Right, I'd forgotten that the parent is needed to `load` the key (forgot until the "This probably also needs to regenerate the parent" part of my comment above)

**Comment by drbild** [comment]
> However, if you don't provide a parent handle to that function, it creates it (or, maybe more precisely, recreates it)

Exactly.   So to use that approach, we would need to be sure to (re)create the same primary key.   Is that safe?   Is there a notion of a "default" primary key? 

That is, the engine uses `TPM2B_PUBLIC_PRIMARY_ECC_TEMPLATE` for the template, a zero-sized `TPM2B_DATA` for the outside info, and zero-count `TPML_PCR_SELECTION` for the creation PCR values.     Is that a safe "default" that we can use in our key generation utility and assume that all TPM engines will support it?

**Comment by drbild** [comment]
> Because the XTT library will need this same functionality, I'm wondering if we should have this key generation stuff in a new repository.

Not advocating for this, but just to discuss.  What about moving the cert generation out of `xtt`?

- `xtt` outputs the public key, private key (possibly TPM-wrapped), and identity (IPv6)
- The client (`enftun`) in our case, has to generate/sign the cert and serialize/store the keys and cert as appropriate.

Advantages include:
- No need for additional library in our ecosystem
- Drop the ASN.1 stuff from `xtt`. `enftun` would just use the openSSL implementations.

Disadvantages include:
- Can openSSL generate certs with the strange lifetimes that we specify?
- Each `xtt` client needs to re-implement the cert creation and storage logic. Or this "new repo" would include the cert generation too.

**Comment by zanebeckwith** [comment]
The `TPM2B_PUBLIC_PRIMARY_ECC_TEMPLATE` is defined by the `tpm2-tss-engine` library ([here](https://github.com/tpm2-software/tpm2-tss-engine/blob/62a60ef9aa08726afa72dcd5361c4089ff78a263/src/tpm2-tss-engine-common.h#L102)). I don't *think* that's a TCG-defined template

That template is exactly the same as what `ecdaa` and `xaptum-tpm` use in their tests (they both have a createprimary-create-load-evict flow, for testing) **except** that they don't include the `NODA` (no dictionary attack protection) flag

This is to say, we *could* try to create exactly the same primary key, but that feels a little dicey. I'm not sure that the other engine creates the exact same primary key (that would be a big pain if not; you could create a key using one engine that wouldn't be loadable using the other engine)

**Comment by zanebeckwith** [comment]
> What about moving the cert generation out of `xtt`

That might be a good idea, and I've always hated having it in `xtt`, but that doesn't save `xtt` from needing this functionality.

The `xtt` handshake still needs to create the key and sign a hash of the handshake.

Or, at the least, even if we say "if using a TPM key, you have to create that key beforehand and provide the handle to `xtt` before the handshake, `xtt` still needs to be able to create a signature

**Comment by drbild** [comment]
Resolved by #128
